pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/MaxAggregationFunction.java

```diff
@@ -54,13 +54,39 @@ public GroupByResultHolder createGroupByResultHolder(int initialCapacity, int ma
   @Override
   public void aggregate(int length, AggregationResultHolder aggregationResultHolder,
       Map<ExpressionContext, BlockValSet> blockValSetMap) {
-    double[] valueArray = blockValSetMap.get(_expression).getDoubleValuesSV();
     double max = aggregationResultHolder.getDoubleResult();
-    for (int i = 0; i < length; i++) {
-      double value = valueArray[i];
-      if (value > max) {
-        max = value;
+    BlockValSet blockValSet = blockValSetMap.get(_expression);
```

Comment:
I understand how delaying conversion to double will help with heap usage. Couple of questions on performance:

- Now every aggregate() call has to execute switch condition once. Will there be any perf penalty for this ?
- Why does conversion to double[] prevent auto-vectorizatonn ?

Reply:
You can see some numbers on #7920 (read longs as longs vs convert to double) but I’ll pull out some disassembly to demonstrate, as well as attach some numbers here.

Reply:
Regarding the switch statement, there is one per block, so it’s cost is amortised (just like all the virtual calls we do)

Reply:
@siddharthteotia I've added some rationale and numbers to the PR description, please take a look.

Reply:
Thanks for sharing perf numbers, @richardstartin. I was also curious to see the disassembled code as you said auto-vectorization probability increases with continuing to treat as long. Please share if you have. 

I am good with the PR. Please also check-in the benchmark

Reply:
The vectorized method is Copy::conjoint_swap, as mentioned above, which hsdis does not dissemble.